src/qat/model/device.py

```diff
@@ -1,118 +1,272 @@
 from __future__ import annotations
 
+import uuid
 from typing import Optional
 
-from pydantic import Field
+import numpy as np
+from pydantic import Field, model_validator
+from pydantic_core import core_schema
 
-from qat.model.component import Component
-from qat.model.serialisation import Ref, RefDict, RefList
-from qat.purr.compiler.devices import ChannelType
+from qat.utils.pydantic import WarnOnExtraFieldsModel
+
+QubitId = int
+
+
+class CalibratablePositiveFloat(float):
+    @classmethod
+    def validate(cls, v):
+        if not np.isnan(v) and v < 0.0:
+            raise ValueError("value must be >= 0")
+        return cls(v)
+
+    @classmethod
+    def __get_pydantic_core_schema__(cls, source_type, handler) -> core_schema.CoreSchema:
+        return core_schema.no_info_plain_validator_function(cls.validate)
+
+    def __eq__(self, other: CalibratablePositiveFloat):
+        if np.isnan(self) and np.isnan(other):
+            return True
+        else:
+            return self.__eq__(other)
+
+    def __neq__(self, other: CalibratablePositiveFloat):
+        return not self.__eq__(other)
+
+
+class Component(WarnOnExtraFieldsModel):
+    """
+    Base class for any logical object which can act as a target of a quantum action
+    - a Qubit or various channels for a simple example.
+
+    :param uuid: The unique string representation of the component.
+    """
+
+    uuid: str = Field(default_factory=lambda: str(uuid.uuid4()), frozen=True)
+
+    def __init__(self, **data):
```

Comment:
trivial constructor, unless you (plan to) have attributes to define (uuid ??) then it could be removed